homeassistant/helpers/template.py

```diff
@@ -2048,6 +2050,61 @@ def __bool__(self) -> bool:
         return super().__bool__()
 
 
+async def async_materialize_hass_loader(hass: HomeAssistant) -> None:
+    """Load all custom jinja into memory."""
+    return await hass.async_add_executor_job(_materialize_hass_loader, hass)
+
+
+def _materialize_hass_loader(hass: HomeAssistant) -> None:
+    result = {}
+    jinja_path = hass.config.path("custom_jinja")
+    all_files = [
+        item
+        for item in pathlib.Path(jinja_path).rglob("*")
+        if item.is_file() and item.stat().st_size < 1024 * 1024
+    ]
+    for file in all_files:
+        content = file.open().read()
+        path = str(file.relative_to(jinja_path))
+        result[path] = content
+
+    _get_hass_loader(hass).sources = result
+
+
+def _get_hass_loader(hass: HomeAssistant) -> HassLoader:
+    if _HASS_LOADER not in hass.data:
+        hass.data[_HASS_LOADER] = HassLoader({})
```

Comment:
Use the `singleton` decorator from `helpers/singleton.py` that will do this automatically for you.